[kubernetes] add prometheus receiver docs (#3114)
open-telemetry/opentelemetry.io: content/en/docs/kubernetes/collector/components.md

title: Important Components for Kubernetes
linkTitle: Components
# prettier-ignore

> cSpell:ignore: containerd crio filelog gotime horizontalpodautoscalers iostream k8sattributes kubelet kubeletstats logtag replicasets replicationcontrollers resourcequotas statefulsets varlibdockercontainers varlogpods alertmanagers

Reviewer: @TylerHelmuth - please re-run `npm run fix:dict` and commit changes.

- The PR will fail otherwise once the following lands: #3116

  cluster-level metrics and entity events.
- [Kubernetes Objects Receiver](#kubernetes-objects-receiver): collects objects,
  such as events, from the Kubernetes API server.

> - [Prometheus Receiver](#kubernetes-objects-receiver): receives metrics in

Reviewer: Copy-paste error:
```suggestion
- [Prometheus Receiver](#prometheus-receiver): receives metrics in
```

In know that we wouldn't get the same effect, but the page has an autogenerated TOC:

> <img width="1655" alt="Screen Shot 2023-07-31 at 17 04 35" src="https://github.com/open-telemetry/opentelemetry.io/assets/4140793/5ae9397d-2298-437e-b3fa-b28e8f8875ca">

Might it be worth dropping this autogenerated one?

Actually, if ever this page is split into separate pages, then we'll get the equivalent of the manually added TOC above automatically for the landing page -- i.e., including the description.

    name: otel-collector-opentelemetry-collector
    namespace: default
```

## Prometheus Receiver

> Prometheus is a common format for Kubernetes metrics. The Prometheus receiver is

Reviewer 2: By kubernetes metrics, do you mean metrics from kubernetes components (e.g. kubelet), or metrics from applications running on k8s?  If the former, https://kubernetes.io/docs/reference/instrumentation/metrics/ may be helpful to link to.

Author: I'm not a Prometheus user myself, but my understanding was that both sources of metrics could be collected as Prometheus metrics.  Is that accurate?

Reviewer 2: Yes, that is accurate.

Author: I'll call them both out specifically

meant to be a minimal drop-in replacement for the collection of those metrics.

> It supports the full set of Prometheus scrape_config options.

Reviewer: ```suggestion
It supports the full set of Prometheus `scrape_config` options.
```

Or maybe link to https://prometheus.io/docs/prometheus/1.8/configuration/configuration/#scrape_config?

Author: Linked

There are a few advanced Prometheus features that the receiver does not support.
The receiver will return an error if the configuration YAML/code contains any of
the following:

> - alert_config.alertmanagers
> - alert_config.relabel_configs
> - remote_read
> - remote_write
> - rule_files

Reviewer: ```suggestion
- `alert_config.alertmanagers`
- `alert_config.relabel_configs`
- `remote_read`
- `remote_write`
- `rule_files`
```

For specific configuration details, see

> [Kubernetes Objects Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver).

Reviewer 2: ```suggestion
[Prometheus Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver).
```
typo?

Author: ya

    name: otel-collector-opentelemetry-collector
    namespace: default
```

## Prometheus Receiver

Prometheus is a common metrics format for both Kubernetes and services running
on Kubernetes. The Prometheus receiver is meant to be a minimal drop-in
replacement for the collection of those metrics. It supports the full set of

> Prometheus scrape_config options.

Reviewer: ```suggestion
Prometheus `scrape_config` options.
```

Or link to https://prometheus.io/docs/prometheus/1.8/configuration/configuration/#scrape_config?

    name: otel-collector-opentelemetry-collector
    namespace: default
```

## Prometheus Receiver

| Deployment Pattern   | Usable |
| -------------------- | ------ |
| DaemonSet (Agent)    | Yes    |
| Deployment (Gateway) | Yes    |
| Sidecar              | No     |

> Prometheus is a common metrics format for both Kubernetes and services running
> on Kubernetes. The Prometheus receiver is meant to be a minimal drop-in
> replacement for the collection of those metrics. It supports the full set of
> Prometheus
> [`scrape_config` options](https://prometheus.io/docs/prometheus/1.8/configuration/configuration/#scrape_config).

Reviewer 3: ```suggestion
Prometheus is a common metrics format for both Kubernetes and services running
on Kubernetes. The Prometheus receiver is a minimal drop-in
replacement for the collection of those metrics. It supports the full set of
Prometheus
[`scrape_config` options](https://prometheus.io/docs/prometheus/1.8/configuration/configuration/#scrape_config).
```

There are a few advanced Prometheus features that the receiver does not support.

> The receiver will return an error if the configuration YAML/code contains any of

Reviewer 3: ```suggestion
The receiver returns an error if the configuration YAML/code contains any of
```

the following:

- `alert_config.alertmanagers`
- `alert_config.relabel_configs`
- `remote_read`
- `remote_write`
- `rule_files`

For specific configuration details, see
[Prometheus Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver).

The Prometheus receiver is
[Stateful](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/standard-warnings.md#statefulness),
which means there are important details to consider when using it:

> - The collector cannot auto-scale the scraping when multiple replicas of the
>   collector are run.
> - When running multiple replicas of the collector with the same config, it will
>   scrape the targets multiple times.
> - Users need to configure each replica with different scraping configuration if
>   they want to manually shard the scraping.

Reviewer 3: ```suggestion
- The collector cannot auto-scale the scraping process when multiple replicas
   of the collector are run.
- When running multiple replicas of the collector with the same config, it will
  scrape the targets multiple times.
- Users need to configure each replica with a different scraping configuration if
  they want to manually shard the scraping process.
```

> To make configuring the Prometheus receiver easier the OpenTelemetry Operator
> includes an optional component called the
> [Target Allocator](../../operator/target-allocator). This component can be used
> to tell a collector which Prometheus endpoints it should scrape.

Reviewer 3: ```suggestion
To make configuring the Prometheus receiver easier, the OpenTelemetry Operator
includes an optional component called the
[Target Allocator](../../operator/target-allocator). This component can be used
to tell a collector which Prometheus endpoints it should scrape.
```
